File: orchestra/ports.py
"""The extension contracts.

Adapters do transport and normalization. The engine owns scheduling policy.
Channels observe the completed tick and cannot influence its decisions.
"""

from __future__ import annotations

from pathlib import Path
from typing import Mapping, Protocol, Sequence

from .models import (
    HealthReport,
    RunSnapshot,
    StartResult,
    TaskObservation,
    TickReport,
    WorkItem,
)


class Board(Protocol):
    name: str

    def fetch(self) -> Sequence[WorkItem]: ...

    def health(self) -> HealthReport: ...


class ActivitySource(Protocol):
    name: str

    def observe(
        self, items: Sequence[WorkItem]
    ) -> Mapping[str, TaskObservation]: ...

    def health(self) -> HealthReport: ...


class AgentRuntime(Protocol):
    name: str

    def is_alive(self, task_dir: Path) -> bool: ...

    def start(
        self,
        task_dir: Path,
        prompt: str,
        resume: bool,
        sequence: int,
    ) -> StartResult: ...

    def stop(self, snapshot: RunSnapshot, *, force: bool = False) -> None: ...

    def health(self) -> HealthReport: ...


class Channel(Protocol):
    name: str

    def publish(self, report: TickReport) -> None: ...

    def health(self) -> HealthReport: ...

File: orchestra/ports.ts
import type {
  HealthReport,
  RunSnapshot,
  StartResult,
  TaskObservation,
  TickReport,
  WorkItem,
} from "./models.js";

/** Adapters may be synchronous fakes or asynchronous transports. */
export type Awaitable<T> = T | Promise<T>;
export interface HealthComponent {
  readonly name: string;
  health(): Awaitable<HealthReport>;
}
export interface Board extends HealthComponent {
  fetch(): Awaitable<readonly WorkItem[]>;
}
export interface ActivitySource extends HealthComponent {
  observe(
    items: readonly WorkItem[],
  ): Awaitable<Readonly<Record<string, TaskObservation>>>;
}
export interface AgentRuntime extends HealthComponent {
  isAlive(taskDir: string): Awaitable<boolean>;
  start(
    taskDir: string,
    prompt: string,
    resume: boolean,
    sequence: number,
  ): Awaitable<StartResult>;
  stop(snapshot: RunSnapshot, options?: { force?: boolean }): Awaitable<void>;
}
export interface Channel extends HealthComponent {
  publish(report: TickReport): Awaitable<void>;
}
